### DISCUSSION TITLE
Problems to iterate a model in OMPython that works in OMEdit smoothly

### DISCUSSION THREAD

**Discussion by kmerbcn** [OpenModelica/OpenModelica#12658] [General]
Hello,

I have a strange problem. I have the following model that works perfectly in OMEdit.

![image](https://github.com/OpenModelica/OpenModelica/assets/62482095/934e68e5-68f1-4bf3-bd92-c69e56aa96c3)

As soon as I call this model in OMPython and I build the system. I have got the following error :

![image](https://github.com/OpenModelica/OpenModelica/assets/62482095/06fec1e3-a141-4e38-b2f0-1d7805900c13)

I am not sure  why I have this error. I know what component from my modelica model makes it. As soon as I remove the diodes, the model disappears and I can compute. Nevertheless, I need these diodes otherwise solutions are not totally realistic and Ideal diodes provide chattering. What command in OMPython am I missing to overcome this jacobian problem?

I hope I gave enough details.

**Comment by casella** [comment]
@kmerbcn this is strange. Could you open a proper issue about it?

**Comment by casella** [comment]
It would be good if you could also share the model, so we can reproduce the error.

**Comment by kmerbcn** [comment]
[Problems.zip](https://github.com/user-attachments/files/16039585/Problems.zip)
 Hello Francesco
I attach the models. Inside of the zip file you have the ressonant converter together with python script. You need to change the folders to run it. 

If you can manage to find the problem, I would be happy to know how you were debugging it. 

Keep in touch master!

**Comment by arun3688** [comment]
@kmerbcn I ran your model in OMEdit and i get the following error 

```
The initialization finished successfully without homotopy method.'

Chattering detected around time 5.010000001e-08..9.80501000098e-05 (100 state events in a row with a total time delta less than the step size 0.002). This can be a performance bottleneck. Use -lv LOG_EVENTS for more information. The zero-crossing was: PWM.deadTime1.continuousClock.y >= PWM.deadTime1.tswitch + 0.5 * PWM.deadTime1.Tdead and PWM.deadTime1.continuousClock.y < PWM.deadTime1.tswitch + 0.5 * (PWM.deadTime1.Ts - PWM.deadTime1.Tdead)
```

Which Openmodelica version you are running ?

**Reply by casella** [reply]
Same for me :)

**Reply by kmerbcn** [reply]
Strange :S

**Reply by casella** [reply]
@kmerbcn I suspect your model has a [sliding mode](https://en.wikipedia.org/wiki/Sliding_mode_control), i.e. it has a discontinuous behaviour that causes the solution to oscillate (in principle) at infinite frequency while moving along a certain surface in the state space, which is where the discontinuity happens.

If you run the model with LOG_EVENTS as suggested by the error message

![immagine](https://github.com/OpenModelica/OpenModelica/assets/4233148/e0c839b8-36ac-4ea3-8cc7-e30f1c5ae52e)

you'll see this chain of events repeating endlessly:
```
state event at time=1.6324218037e-06
[4] not (-Lc2.i) > gThreshold.threshold
[3] (-Lc2.i) > gThreshold.threshold
state event at time=3.4501000007e-06
[14] PWM.deadTime.continuousClock.y >= PWM.deadTime.tswitch + 0.5 * PWM.deadTime.Tdead and PWM.deadTime.continuousClock.y < PWM.deadTime.tswitch + 0.5 * (PWM.deadTime.Ts - PWM.deadTime.Tdead)
[13] PWM.deadTime1.continuousClock.y >= PWM.deadTime1.tswitch + 0.5 * PWM.deadTime1.Tdead and PWM.deadTime1.continuousClock.y < PWM.deadTime1.tswitch + 0.5 * (PWM.deadTime1.Ts - PWM.deadTime1.Tdead)
state event at time=3.57291666737e-06
[11] PWM.trapezoid.y >= Vref.k
state event at time=3.62301666773e-06
[14] PWM.deadTime.continuousClock.y >= PWM.deadTime.tswitch + 0.5 * PWM.deadTime.Tdead and PWM.deadTime.continuousClock.y < PWM.deadTime.tswitch + 0.5 * (PWM.deadTime.Ts - PWM.deadTime.Tdead)
state event at time=3.62302786283e-06
```
As I understand, as soon as S3 and S4 are switched, the derivative of the capacitor voltage changes the sign, causing the threshold to be immediately switched again, leading to infinite frequency oscillations.

I real life there will be some delay or dynamics that makes these oscillations happen with a finite frequency, but you have apparently abstracted that away in your model. You should add some hysteresis or delay on the feedback loop to prevent this, e.g. by adding a delay or low-pass filter block at the `Isense` output.

**Reply by kmerbcn** [reply]
What Ill do is to sample the information of the current in a way it should be done in an embedded system. 
The information of the current is to be sampled by an ADC in a microcontroller. Then, check for change of sign and eventually trigger one of the switches. I think, that might solve the problem.

Br,

**Reply by casella** [reply]
You definitely need to represent some kind of delay in the model. Either you use Clocked blocks with regular sampling (and a one-sample delay between input and output), or a pure time delay like Modelica.Blocks.Nonlinear.FixedDelay (BTW, Fixed Delay is linear, it ended up in Nonlinear to mimick Simulink, which put it there 30+ years ago...).

Depending on how you do it, your simulation could go from slow to terribly slow. But at least it won't get stuck as it does now.

**Reply by kmerbcn** [reply]
It works perfect now

**Reply by kmerbcn** [reply]
For example. I moved from Simulink to Modelica after using Altair Activate. Activate derives from Modelica and I knew about the existance of Modelica there. I remember they had a block "Break algebraic loop". I suppose it must be associated to that delay.
Does it have something to do the fixed delay with the break algebraic loop?

**Reply by casella** [reply]
Not really. Modelica tools solve algebraic loops in the proper way, i.e. using implicit equation solvers. The problem you have here is different: you have a discontinuous dynamical system for which a classical solution of the ODEs does not exist, hence the ODE solver gets in trouble.

Something like der(x) = -sign(x). The sufficient conditions for Cauchy's theorem of local existence and uniqueness of the solution are not satisfied, for that the RHS should be Lipschitz continuous, and sign(x) obviously isn'.

**Reply by kmerbcn** [reply]
Thanks for the clarification Francesco.
At the end the problem is instrinsical to the dynamics of the non-linear system and not on the discretization scheme of the ODE.
Thanks!

**Reply by kmerbcn** [reply]
Hello Francesco!

I have a small technical question concerning the discussion of yesterday. Coming back the topic about the existance of solution of an ODE and the problems dealing with chattering, I would like to know why a delay improves Liptschitz conditions.
In my mind (correct me if i am wrong), discontinuity is still there. Isn't it?

Br,

Javier Martinez

**Reply by casella** [reply]
Those are sufficient conditions, but by no means necessary :)

A delay makes the infinite switching frequency finite. Another approach used in sliding mode control is to somehow regularize the discontinuity. But that is not how switching systems are handled. Switching systems often use "sliding modes" with high-frequency switching, the point is that the factor that determines the high (but not infinite) frequency must be modelled explicitly, albeit approximately if you want to do system-level simulation.

**Reply by kmerbcn** [reply]
Thanks for the clarification master. Are these topics in the book of continouous time systems?

**Reply by casella** [reply]
Not really. It discusses the proper handling of discontinuous systems by event handling, but there is no discussion of sliding modes and chattering. I learned about sliding modes for my master's thesis, see http://dx.doi.org/10.1016/0967-0661(96)00070-6

**Reply by kmerbcn** [reply]
Nice work. I will fetch some ideas since I worked briefly with slidding models in power converters. I think is a very interesting topic altogether.

**Comment by kmerbcn** [comment]
Good morning Francesco!
I had time this week end to go through the paper you mentioned. I liked the aprroach to smooth the slidding control.
As I mentioned before, I did some trials to use slidding control on power converters.
I found the approach nice and I would like to know if there is a way to implement equation (8) of your paper in Modelica. As I see, you can do it by writing your time solver. But in dassl is there a way to implement equation 8?

**Reply by casella** [reply]
Do you mean this equation: 

![immagine](https://github.com/user-attachments/assets/e488cd8d-c4db-4454-808d-74f03fe7abd0)

This is just a result of the analysis, it's not the control law.

**Reply by kmerbcn** [reply]
I went thorugh the paper again and implemented the control law based on Fig. 6 . The computing time among the classical slidding rule and the new rule is way better. In the past as I mentioned, I did some slidding control of power converters and I found that the switching times were two high if you implement slidding theory. My experience in Modelica was that error was zero because switching tended to infinte. Another paper, taken from IEEE, controlled the switching time and also the error. Perhaps I am saying something stupid, but is this ratio between switching time and steady state error has to do something with Filipov dynamical systems? (Sorry for my ignorance)

Have a good summer Francesco!
